Ill 151 admin can update user role

JonathanGian opened the pull request:
![image](https://github.com/user-attachments/assets/c12e229a-ea88-4fb0-b914-f6f54dd7cc43)
This pull request introduces significant updates to the user management functionality, including new APIs, components, and routing changes, as well as minor UI improvements. The most notable changes focus on enhancing the ability to manage user roles and statuses, adding a dedicated "Manage Users" page, and improving navigation for admin users.

### User Management Enhancements:
* Updated the `updateUserStatus` API in `frontend/src/api/users.ts` to support additional statuses (`'deactivated'` and `'active'`) and introduced new APIs for updating user roles, promoting users to admin, and approving unapproved users.
* Added a new `ManageUsers` component (`frontend/src/components/Admin/ManageUsers.tsx`) that includes a data grid for filtering, viewing, and updating user roles and statuses. This component supports tab-based filtering and integrates with Redux for state management.
* Introduced a new `updateUserRole` thunk in `frontend/src/slices/usersSlice.ts` to handle role updates and added corresponding Redux state management logic. [[1]](diffhunk://#diff-c2d7ec6364fd86cf9dfe94c25c562f5c3fc47d03891fdc03befc033f6ca389c2R63-R74) [[2]](diffhunk://#diff-c2d7ec6364fd86cf9dfe94c25c562f5c3fc47d03891fdc03befc033f6ca389c2R162-R182)

### Routing and Navigation Updates:
* Added a route for the "Manage Users" page (`/admin/users`) in `frontend/src/router/router.tsx`.
* Updated the "Manage Users" button in the `AdminDashboard` component to link to the new "Manage Users" page.
* Added a "Dashboard" link to the `PersonMenu` component for easier navigation to the admin dashboard.

### Minor UI and Code Improvements:
* Adjusted the `AdminDashboard` component's dependency arrays and memoized values for better performance. [[1]](diffhunk://#diff-1c8194e90841b07f5be90b612aa54d1c1501eea5cf87d867ae17b671e2ebc12aL95-R94) [[2]](diffhunk://#diff-1c8194e90841b07f5be90b612aa54d1c1501eea5cf87d867ae17b671e2ebc12aL122)
* Improved `DataGrid` styling in the `ManageUsers` component for a more polished user experience.

athinakantis reviewed (changes requested):
### **What I tested**
As an admin I tried the following things
- Updating a users status: 
This does work, however after initially changing the users status, the table refreshes and doesn't show the new changes. At the moment I have to refresh to see if it worked or not.
- Updating a users role
- This does not seem to work for me. I tried updating a user from Unapproved -> User without success


### **Other improvements**
#### UI Improvements
- The grey used in the manage users table is too dark, it gets distracting. Considering changing it to `background.verylightgrey` or similar
- Theres no need for a purple border between the rows as the white/grey already separates the rows from each other
- The heading should be of variant `heading_secondary_bold` to align with all our other pages
- There is currently a warning in the console with the following message:
```js
users.ts:29 MUI X: The page size `100` is not present in the `pageSizeOptions`.
Add it to show the pagination select. Error Component Stack
    at ManageUsers (ManageUsers.tsx:27:20)
    at Root (Root.tsx:13:24)
    at AuthProvider (AuthProvider.tsx:18:32)
    at ThemeProvider (ThemeProvider.tsx:9:33)

```

#### **User Experience**
- We should consider adding a notification. `_user_name_ has been approved!`. After approving a user I want some kind of confirmation, did it go through? 

#### **Functionality**
- For some reason the users table shows rows where a user with no name or email.
- The `Pending Approvals` tab doesn't show only users with an `Unapproved` status
- Active tab doesn't show anything (Maybe this will be done later?)
- `Pending Approvals` doesn't show only users with an `Unapproved` role
- New users should have a default "Active" status (Only deactivated if an admin deems the account/person not fit to make any sort of bookings or similar)

JonathanGian commented:
> ### **What I tested**
> As an admin I tried the following things
> 
> * Updating a users status:
>   This does work, however after initially changing the users status, the table refreshes and doesn't show the new changes. At the moment I have to refresh to see if it worked or not.

- This was an issue in the backend. I put the wrong checks on the function, It was looking for only a Head Admin to change status but its fixed now.

> * Updating a users role
>   This does not seem to work for me. I tried updating a user from Unapproved -> User without success

- There were quite a few bugs in the component. I ended up scrapping it and starting again but it should now work for both Admins and Head Admins. (Head Admins can change to any role and Admins can only upgrade Unapproved -> User

> 
> ### **Other improvements**
> #### UI Improvements
> * The grey used in the manage users table is too dark, it gets distracting. Considering changing it to `background.verylightgrey` or similar
> * Theres no need for a purple border between the rows as the white/grey already separates the rows from each other
> * The heading should be of variant `heading_secondary_bold` to align with all our other pages
> * There is currently a warning in the console with the following message:
> 

- I scrapped the whole thing and left the styling out so someone else can do it.

> ```js
> users.ts:29 MUI X: The page size `100` is not present in the `pageSizeOptions`.
> Add it to show the pagination select. Error Component Stack
>     at ManageUsers (ManageUsers.tsx:27:20)
>     at Root (Root.tsx:13:24)
>     at AuthProvider (AuthProvider.tsx:18:32)
>     at ThemeProvider (ThemeProvider.tsx:9:33)
> ```
> 

- Forgot to add all page amounts to the component. Fixed now.

> * There should be a max width for the page similar to our other pages. It should not span the entire width of the page
> 
> #### **User Experience**
> * We should consider adding a notification. `_user_name_ has been approved!`. After approving a user I want some kind of confirmation, did it go through?
> 
It was a good idea and is now in the component.

> #### **Functionality**
> * For some reason the users table shows rows where a user with no name or email.

- This is because we have at least one user in the database that I havent given a display_name to and I signed up with my phone number so the information is empty.

> * The `Pending Approvals` tab doesn't show only users with an `Unapproved` role

- Almost overlooked this again when redoing the component but it was a small fix. The letters were in uppercase and the database is lowercase.

> * `Pending Approvals` doesn't show only users with an `Unapproved` role

- Fixed but there is a small issue. Our status defaults to "pending" so we cant have it default ot "active" but i did make it so the tabs make sense. 

```js
  const filtered = users.filter((u) => {
    if (filter === 'ALL') return true;
    // Lists pending users
    if (filter === 'PENDING') return u.user_status === 'pending';
    // Lists all active users(we need to figure out our database structure)
    if (filter === 'ACTIVE') return u.user_status === 'active'|| u.user_status === 'approved'; 
    if (filter === 'DEACTIVATED') return u.user_status === 'deactivated';
    return true;
  });
```

> * New users should have a default "Active" status (Only deactivated if an admin deems the account/person not fit to make any sort of bookings or similar)

- I agree. Ignore what the statuses say for now, I havent had a chance to make policies or anything like that to make the status actually do anything. 
- Right now they are just unused statuses that can conflict with reasonable thinking.
- Also we need to agree together on how the database will work because I still have no idea what kinda policies to make or things to do in Supabase.
- There has been conflicting ideas with how the status/unapproved things should work so we need to come to an agreement together so I or you can make the changes.

JonathanGian commented on backend/src/guards/role.guard.ts:
I didint touch this its just from merging with DEV

JonathanGian commented on backend/src/services/admin.service.ts:
Changed it so both Admins and HeadAdmins can change the role. Then I gaurded it in the frontend so Admins can upgrade Unapproved to Users and Head Admins can change to any role

JonathanGian commented on frontend/src/api/users.ts:
Added all endpoints from backend but didint use most of them

JonathanGian commented on frontend/src/components/Account.tsx:
There are a lot of these single changes because I went through almost all app errors to either fix or silence them

JonathanGian commented on frontend/src/components/Admin/AdminBookings/AdminBookings.tsx:
Removed the old Snackbar and updated to Nikitas

JonathanGian commented on frontend/src/components/Admin/AdminDashboard.tsx:
Fixed errors and also updated the link so it directs to the proper page

JonathanGian commented on frontend/src/components/Admin/ManageUsers.tsx:
Scapped the old one and started over again. Its mostly the same

JonathanGian commented on frontend/src/context/AuthProvider.tsx:
The old role system broke and I have no idea what broke it yet. But this will work for now and in some ways seems more reliable. There isnt that issue where role is undefined for so long when you refresh

JonathanGian replied:
Also, I just thought of a way to maybe fix the issue of the role taking a few seconds to be defined. We could set some timers that make the page load just a bit longer before deciding which role the user has. That way we dont have the situation where a normal User might see a glimpse of the Admin pages before being redirected. Seems like it will work in theory 😄

JonathanGian commented on frontend/src/types/users.type.ts:
made type of exactly what the backend gives us

JonathanGian commented on frontend/src/utility/SnackBar.tsx:
Removed my old Snackbar
